### PR TITLE
J3: self-distillation during training (EMA-teacher + stochastic-depth)

### PR DESCRIPTION
**Pull request by gmboquet**
## Stacking

Stacked on top of `coarsening-operator` (G3, PR #151) -- this branch is based on that branch's tip because the acceptance test below calls G3's `coarsen()` directly. Please merge #151 first (or review this diff as G3 + this commit).

## What

Roadmap J3: EMA-teacher consistency and stochastic-depth targets applied *during* normal training, inside `mixle/models/self_distillation.py`:

- `EMATeacher(model, decay)` -- maintains a mean-teacher EMA copy of a model's own weights (`teacher = decay*teacher + (1-decay)*student`), updated via `.update(student_model)`; `.forward`/`.predict` run the frozen teacher. Handles `CausalLM`'s weight-tied `tok`/`head` embedding correctly (dedupes by storage identity so a tied tensor isn't EMA-updated twice per step).
- `consistency_loss(student_output, teacher_output, mode="mse"|"kl")` -- the self-distillation loss term.
- `stochastic_depth_forward(model, x, drop_prob)` -- runs the same input at full depth and with a random subset of blocks skipped (standard stochastic depth / drop-path), returning `(full_output, partial_output)` for consistency.
- `train_with_self_distillation(model, data, steps, ema_decay, drop_prob, consistency_weight, ...)` -- wires both pressures into ordinary next-token cross-entropy training as extra additive loss terms, mirroring `GradLeaf`'s loss-hook composition pattern and `language_model.py`'s own training loop conventions (no new trainer abstraction). `data` is shaped exactly like `mixle.data.stream_token_source.stream_token_source`'s `(context, next_token)` batches.

Searched first for existing EMA / stochastic-depth machinery (`grep -rln "exponential.moving.average\|ema_decay\|class EMA"` and `"stochastic.depth\|drop.path\|droppath"` across `mixle/`) -- neither exists yet, so both are implemented fresh here.

## Acceptance criterion -- substituting G3's `coarsen()` for J2's not-yet-built "ladder"

J2 (checkpoint-compression ladder) and F1 (distributed trainer) don't exist yet. Per the roadmap note, the acceptance test (`mixle/tests/self_distillation_test.py::J3CompressibilityAcceptanceTest`) substitutes G3's already-built `coarsen()` (PR #151) as the compressibility proxy: a small `CausalLM` is trained two ways for the *same* number of steps on the *same* synthetic data -- (a) plain cross-entropy, (b) `train_with_self_distillation` -- then both checkpoints are run through `coarsen()` at the *same* divergence budget and trust region.

Measured result (printed by the test, `-n0 -m ""`):

```
[J3 acceptance] plain: accepted_pairs=3 total_kl=2.063826 depth=3
[J3 acceptance] j3:    accepted_pairs=3 total_kl=1.001612 depth=3
```

Same architecture (6 -> 3 layers, full 2x depth cut accepted for both), same budget/trust region: the J3-trained checkpoint's accumulated divergence is ~2x lower for the identical depth cut -- the causal claim ("self-consistency pressure -> flatter, more redundant, more coarsen-friendly solutions") holds in this measured run.

## Tests

`mixle/tests/self_distillation_test.py` (6 tests, all pass):
1. `EMATeacherCorrectnessTest` -- EMA math checked numerically against a hand-computed average across 3 update steps (plus a frozen/forward sanity check).
2. `StochasticDepthCorrectnessTest` -- `drop_prob=0` gives byte-identical full/partial outputs; `drop_prob=0.9` gives a measurably different partial output; `consistency_loss` sanity (zero for identical inputs, positive otherwise).
3. `J3CompressibilityAcceptanceTest` -- the core acceptance criterion above.

No regressions: `mixle/tests/coarsening_test.py` (10 tests) still passes unchanged.

## Workflow

- `ruff check --fix` / `ruff format` clean on both new files.
- `pytest mixle/tests/self_distillation_test.py mixle/tests/coarsening_test.py -n0 -m ""` -- 16/16 pass.

### CODE REVIEW

**Comment by gmboquet** [comment]
Relanded as #220 (merged).
